### ISSUE TITLE
OptimalTTT learner crashes in some configurations

### ISSUE THREAD

**Issue by stateMachinist** [LearnLib/learnlib#135]
I've performed some experiments including the OptimalTTT learner and noticed that it crashes in some configurations. The following code triggers an internal assertion in OptimalTTT, using the latest official LearnLib release 0.17 with Java 17.

```
import de.learnlib.algorithm.LearningAlgorithm.MealyLearner;
import de.learnlib.algorithm.oml.ttt.mealy.OptimalTTTMealy;
import de.learnlib.oracle.EquivalenceOracle.MealyEquivalenceOracle;
import de.learnlib.oracle.equivalence.MealyRandomWpMethodEQOracle;
import de.learnlib.oracle.membership.MealySimulatorOracle;
import net.automatalib.automaton.transducer.CompactMealy;
import net.automatalib.serialization.dot.DOTParsers;
import net.automatalib.util.automaton.Automata;

import java.io.IOException;
import java.util.*;

public class Main {

    public static void main(String[] args) {
        final var model = loadModel("mosquitto__two_client_will_retain.dot");
        final var mqo = new MealySimulatorOracle<>(model);
        final var learner = new OptimalTTTMealy<>(model.getInputAlphabet(), mqo);
        final int eqoSeed = 36;
        final var eqo = new MealyRandomWpMethodEQOracle<>(mqo, 0, 4, 1000000, new Random(eqoSeed), 1);
        learnToExhaustion(model, learner, eqo);
    }

    private static CompactMealy<String, String> loadModel(String name) {
        try {
            final var is = Benchmark.class.getClassLoader().getResourceAsStream(name);
            return DOTParsers.mealy().readModel(is).model;
        } catch (IOException e) {
            throw new RuntimeException(e.getMessage());
        }
    }

    private static void learnToExhaustion(CompactMealy<String, String> model,
                                          MealyLearner<String, String> learner,
                                          MealyEquivalenceOracle<String, String> eqo) {
        learner.startLearning();
        while (true) {
            if (Automata.findSeparatingWord(model, learner.getHypothesisModel(), model.getInputAlphabet()) == null) {
                return;
            }

            // the model is still incomplete
            var cex = eqo.findCounterExample(learner.getHypothesisModel(), model.getInputAlphabet());
            if (cex == null) {
                throw new RuntimeException("counterexample exists, but was not found");
            }

            learner.refineHypothesis(cex);
        }
    }
}
```

Running this code results in the following error:

```
Exception in thread "main" java.lang.AssertionError
	at de.learnlib.algorithm.oml.ttt.AbstractOptimalTTT.longestShortPrefixOf(AbstractOptimalTTT.java:155)
	at de.learnlib.algorithm.oml.ttt.AbstractOptimalTTT.analyzeCounterexample(AbstractOptimalTTT.java:161)
	at de.learnlib.algorithm.oml.ttt.AbstractOptimalTTT.refineWithWitness(AbstractOptimalTTT.java:134)
	at de.learnlib.algorithm.oml.ttt.AbstractOptimalTTT.refineHypothesis(AbstractOptimalTTT.java:87)
	at Main.learnToExhaustion(Main.java:48)
	at Main.main(Main.java:21)
```

The model file is from [here](https://gitlab.com/felixwallner/learn-combinations/-/tree/master/input/mealy/principle/BenchmarkMQTT). The problem only occurs for specific combinations of models and seeds. E.g. when changing the seed to 35, this model is learned successfully and the code terminates without problems. But the same problem also occurs for other models with different seeds.

The problem is obviously related to counterexample processing. I have read the corresponding paper and understood the principle idea underlying the refinement strategy of $L^\lambda$ (btw, my hats off for that finding!). But I am not familiar enough with the mechanical details of this algorithm to identify the source of this issue. Interestingly, the crashes only occured with the tree-based implementation. I'm not sure if the table-based version also has this problem and I simply haven't tried enough seeds, or whether this is specific to the TTT version.

P.S.: Thank you very much for maintaining LearnLib! My research would not be possible without it.

**Comment by mtf90** [comment]
Dear @stateMachinist, I haven't been able to reproduce the issue with 0.17 (mainly because it takes forever to iterate over all counterexamples). However, we have recently (54fc0d2) made some changes to the way the algorithm analyzes counterexamples, which does seem to affect this use-case. With the current development version, the learning process finished relatively quick without any errors.

If you are willing to test the development version, you can do so with our [nightly builds](https://github.com/LearnLib/learnlib/wiki/Using-Development-Versions). However, we are also planning for a new release early January, so the fixes should be available in a stable version soon, too. (Although some refactoring will be necessary as we plan to rename the algorithm from `oml` to `lambda`).

**Comment by stateMachinist** [comment]
Thank you for the helpful response! I have tested the current snapshot, and the issue seems to be resolved. But I find it strange that you could not reproduce the issue. I first noticed the problem when my experiments would not terminate. Apparently, some refinements never completed and the learning loop kept running in circles (until I enabled assertions to crash it). I guess there is some external factor at play here, like some dependency installed with different versions on our systems. Anyway, it's very nice that the problem is solved! Now I can complete the benchmark results for my paper.

Another question: I wanted to use the stable release to guarantee reproducibility of my results when submitting an artifact. In the development branch, classes are sometimes renamed and the code needs to be adjusted for it to compile, and results might slightly change because of changes to the algorithm implementations. When I am using the development version, is there some way to fix it to a specific version within the POM, so I can guarantee it will always run with the same version? I am currently using 0.18.0-SNAPSHOT. But I simply guessed the version name from other POMs I saw, because I could not find a way to list available versions.

**Comment by mtf90** [comment]
> But I find it strange that you could not reproduce the issue.

Well, I aborted the run when nothing happened after 2 hours (with `-ea` as well). Maybe you can log the counterexamples of your failed run and create a more concise test-case that only uses the relevant ones? If it happens to unveil an issue that is still present in the current implementation, we could directly use it as an integration test, too.


> When I am using the development version, is there some way to fix it to a specific version within the POM, so I can guarantee it will always run with the same version?

No, `-SNAPSHOT` versions are always resolved by date and the newest one wins. I see to options as viable:

* If you build your artifact from code as well, you could add LearnLib as a `git submodule` and build it before your own code (you may need to add the development version of AutomataLib as well). The downside is that you need to re-compile both libraries every time that you build your artifact. The upside is that your local build will always be the newest and therefore take precedence over anything that is deployed on the nightly servers. Plus, you are able to select specific checkouts in the submodule (if you want to switch between development versions).
* You can use the `-Pbundles` profile to build a fat jar and include it as a `.jar` instead via Maven to pin a certain build. Depending on your other dependencies and the classes of LearnLib you want to use, you may either want to use the `-with-dependencies.jar` or the stripped one.

**Comment by stateMachinist** [comment]
> Maybe you can log the counterexamples of your failed run and create a more concise test-case that only uses the relevant ones?

Good idea. I am quite busy today, but I will get back to it in the following days.

> No, -SNAPSHOT versions are always resolved by date and the newest one wins.

Hmm, I see. When I started working on a new learning algorithm, initially I used to work within a fully cloned learnlib environment. But I really did not like the build complexity with that, so I decided to simplify my setup and distill my code base down to the essential ingredients. I might try the second option, but otherwise I will simply backport the OptimalTTT implementation to release 0.17 with the relevant patch manually applied.

**Comment by stateMachinist** [comment]
This is the trace of counterexamples up until the assertion error in the code I've posted above, as received by printing the counterexamples before calling _refineHypothesis_:

```
SubscribeC2 ConnectC2 ConnectC1WithWillRetain UnSubScribeC2 DisconnectTCPC1 UnSubScribeC2 ConnectC2 / c1_ConnectionClosed__c2_ConnectionClosed c1_ConnectionClosed__c2_ConnAck c1_ConnAck__Empty Empty__c2_UnSubAck c1_ConnectionClosed__Empty c1_ConnectionClosed__c2_UnSubAck c1_ConnectionClosed__c2_ConnectionClosed
ConnectC1WithWillRetain ConnectC2 SubscribeC2 ConnectC1WithWill ConnectC1WithWillRetain ConnectC1WithWill ConnectC2 ConnectC1WithWill ConnectC1WithWillRetain / c1_ConnAck__c2_ConnectionClosed Empty__c2_ConnAck Empty__c2_SubAck c1_ConnectionClosed__Pub(c2,my_topic,bye) c1_ConnAck__Empty c1_ConnectionClosed__Pub(c2,my_topic,bye) c1_ConnectionClosed__c2_ConnectionClosed c1_ConnAck__c2_ConnectionClosed c1_ConnectionClosed__c2_ConnectionClosed
ConnectC1WithWillRetain ConnectC1WithWill ConnectC1WithWillRetain ConnectC2 ConnectC1WithWillRetain ConnectC1WithWillRetain ConnectC1WithWillRetain / c1_ConnAck__c2_ConnectionClosed c1_ConnectionClosed__c2_ConnectionClosed c1_ConnAck__c2_ConnectionClosed Empty__c2_ConnAck c1_ConnectionClosed__Empty c1_ConnAck__Empty c1_ConnectionClosed__Empty
ConnectC1WithWillRetain ConnectC2 ConnectC1WithWillRetain SubscribeC2 DeleteRetainedC2 DeleteRetainedC2 ConnectC1WithWillRetain DeleteRetainedC2 ConnectC2 ConnectC2 DeleteRetainedC2 UnSubScribeC2 DeleteRetainedC2 UnSubScribeC2 UnSubScribeC2 DisconnectTCPC1 DisconnectTCPC1 DeleteRetainedC2 / c1_ConnAck__c2_ConnectionClosed Empty__c2_ConnAck c1_ConnectionClosed__Empty c1_ConnectionClosed__c2_SubAck__Pub(c2,my_topic,bye) c1_ConnectionClosed__Pub(c2,my_topic,)__c2_PubAck c1_ConnectionClosed__Pub(c2,my_topic,)__c2_PubAck c1_ConnAck__Empty Empty__Pub(c2,my_topic,)__c2_PubAck Empty__c2_ConnectionClosed Empty__c2_ConnAck Empty__c2_PubAck Empty__c2_UnSubAck Empty__c2_PubAck Empty__c2_UnSubAck Empty__c2_UnSubAck c1_ConnectionClosed__Empty c1_ConnectionClosed__Empty c1_ConnectionClosed__c2_PubAck
ConnectC1WithWillRetain ConnectC2 SubscribeC2 DisconnectTCPC1 DisconnectTCPC1 ConnectC2 DisconnectC1 SubscribeC2 DeleteRetainedC2 ConnectC1WithWillRetain ConnectC2 SubscribeC2 ConnectC1WithWillRetain ConnectC1WithWillRetain ConnectC2 ConnectC1WithWillRetain ConnectC1WithWillRetain / c1_ConnAck__c2_ConnectionClosed Empty__c2_ConnAck Empty__c2_SubAck c1_ConnectionClosed__Pub(c2,my_topic,bye) c1_ConnectionClosed__Empty c1_ConnectionClosed__c2_ConnectionClosed c1_ConnectionClosed__c2_ConnectionClosed c1_ConnectionClosed__c2_ConnectionClosed c1_ConnectionClosed__c2_ConnectionClosed c1_ConnAck__c2_ConnectionClosed Empty__c2_ConnAck Empty__c2_SubAck__Pub(c2,my_topic,bye) c1_ConnectionClosed__Pub(c2,my_topic,bye) c1_ConnAck__Empty Empty__c2_ConnectionClosed c1_ConnectionClosed__c2_ConnectionClosed c1_ConnAck__c2_ConnectionClosed
ConnectC2 SubscribeC2 DisconnectTCPC1 ConnectC1WithWill ConnectC2 ConnectC1WithWillRetain UnSubScribeC2 DeleteRetainedC2 ConnectC1WithWillRetain SubscribeC2 SubscribeC2 DisconnectC1 ConnectC1WithWill ConnectC2 SubscribeC2 / c1_ConnectionClosed__c2_ConnAck c1_ConnectionClosed__c2_SubAck c1_ConnectionClosed__Empty c1_ConnAck__Empty Empty__c2_ConnectionClosed c1_ConnectionClosed__c2_ConnectionClosed c1_ConnectionClosed__c2_ConnectionClosed c1_ConnectionClosed__c2_ConnectionClosed c1_ConnAck__c2_ConnectionClosed Empty__c2_ConnectionClosed Empty__c2_ConnectionClosed c1_ConnectionClosed__c2_ConnectionClosed c1_ConnAck__c2_ConnectionClosed Empty__c2_ConnAck Empty__c2_SubAck
ConnectC1WithWillRetain ConnectC1WithWillRetain ConnectC2 ConnectC1WithWill UnSubScribeC2 SubscribeC2 DeleteRetainedC1 ConnectC1WithWillRetain SubscribeC2 DisconnectC1 SubscribeC2 / c1_ConnAck__c2_ConnectionClosed c1_ConnectionClosed__c2_ConnectionClosed c1_ConnectionClosed__c2_ConnAck c1_ConnAck__Empty Empty__c2_UnSubAck Empty__c2_SubAck__Pub(c2,my_topic,bye) c1_PubAck__Pub(c2,my_topic,) c1_ConnectionClosed__Pub(c2,my_topic,bye) c1_ConnectionClosed__c2_SubAck c1_ConnectionClosed__Empty c1_ConnectionClosed__c2_SubAck
```

The format is _input / output_ and symbols are separated by whitespaces. If you need a different data format, let me know.

You've said that iterating through the counterexamples takes over 2 hours for you. To me, this very much sounds like the run got stuck in an endless loop. Have you tried different seeds, for example 35? If that works, then I think the problem might be related to the assertion not triggering when it should. Because when I had the issue with assertions being ignored, the code would also not terminate.

**Comment by mtf90** [comment]
This may have been an error on my side. I think I messed up the run configuration and specified `-ea` as a program flag, not as a VM flag. At least now I was able to reproduce the issue with your original code. However, your extracted counterexample helped very much in constructing a single run to debug, so thank you.

The good news is, the issue should also be fixed by 54fc0d2. The problem was that the old version always answered intermediate counterexamples [with the full input trace](https://github.com/LearnLib/learnlib/blob/learnlib-0.17.0/algorithms/active/oml/src/main/java/de/learnlib/algorithm/oml/ttt/AbstractOptimalTTT.java#L127) even though they may consist of separate prefixes and suffixes. The [current version](https://github.com/LearnLib/learnlib/blob/develop/algorithms/active/lambda/src/main/java/de/learnlib/algorithm/lambda/ttt/AbstractTTTLambda.java#L76) correctly answers only the suffix part. As a result, the algorithm would analyze counterexamples that in reality aren't any counterexample (the outputs only differed in length) and the consistency checks would fail.
